lib/logsource.c

```diff
@@ -242,6 +247,9 @@ log_source_post(LogSource *self, LogMessage *msg)
 
   old_window_size = g_atomic_counter_exchange_and_add(&self->window_size, -1);
 
+  if (G_UNLIKELY(g_atomic_counter_get(&self->window_size) == 0))
+    msg_debug("Source has been suspended", evt_tag_str("group_name", self->options->group_name));
```

Comment:
This atomic operation is in the fast path and is racy. We should check if old_window_size is 1, that avoids both the race and the atomic.

I cant see the context of this change, but wouldnt this be covered by the message in log_source_flow_control_suspend()?